module/VuFind/src/VuFind/Config/Feature/SecretTrait.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/**
+ * Trait to import secret from file rather than a hardcoded config
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Michigan State University 2024.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Robby ROUDON <roudonro@msu.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\Config\Feature;
+
+use Laminas\Config\Config;
+
+/**
+ * Trait to import secret from file rather than a hardcoded config
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Robby ROUDON <roudonro@msu.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+trait SecretTrait
+{
+    /**
+     * Load a secret value from the specified configuration and key.
+     * Will look for a _file-suffixed version of the key first,
+     * and load the data from a separate file if configured to do so.
+     *
+     * @param Config|array|null $config The config to read from
+     * @param string            $key    The key to retrieve
+     *
+     * @return string|null
+     */
+    protected function getSecretFromConfig(Config|array|null $config, string $key): ?string
+    {
+        if ($config === null) {
+            return null;
+        }
+        if ($config instanceof Config) {
+            $config = $config->toArray();
+        }
+        if (isset($config[$key . '_file'])) {
+            if (is_readable($config[$key . '_file'])) {
+                $value = file_get_contents($config[$key . '_file']);
+                return trim($value);
+            } elseif (method_exists($this, 'logWarning')) {
+                $this->logWarning(
+                    'The secret file (' . $config[$key . '_file'] . ')' .
+                    ' for secret ' . $key . ' doesn\'t exist or is not readable.'
+                );
+            } else {
+                error_log(
+                    'The secret file (' . $config[$key . '_file'] . ')' .
+                    ' for secret ' . $key . ' doesn\'t exist or is not readable'
+                );
+            }
+        }
+        if (isset($config[$key])) {
+            return trim($config[$key]);
```

Comment:
I still don't feel comfortable that this trims spaces, particularly because spaces were previously left alone.

Reply:
Would you feel better with something like `rtrim($config[$key], "\n")`? The main use case is that trailing line breaks can mess things up (especially since some editors like nano force a trailing line break), so I think we should do something to compensate for that problem at least.

Reply:
Also, let's try to remember that any further changes we make to logic here should be matched in the SolrMarc Java code for consistency...

Reply:
Yes, for earlier case where the value is read from the secret file. Here with the single config value I don't see when the trim would be necessary if it wasn't necessary earlier either.

Reply:
Yes, that makes sense to me!

Reply:
I removed it, is it good with you?

Reply:
Looks reasonable to me -- at least, the PHP and Java code are now consistent with one another, which is good. Whether we need to worry about leading or trailing spaces in secret values is a separate question, but I think that's unlikely enough that I'm willing to live with a simple trim() for now unless @EreMaijala feels it's worth doing something more complex.

Reply:
I started thinking that we could perhaps want to use some format that makes this all clear for secret files. But I suppose we can later add e.g. a check for .json in file name and decode the contents accordingly.

Reply:
I could also imagine having a single shared secrets file with different keys in it, with syntax like `filename.json:key`. Lots of possibilities, but I agree that it makes sense to merge this as-is for now and await further use cases before going to deep into the engineering. :-)

Reply:
We use docker secrets and the file contains only the value without any new line

Reply:
Yes, should be fine for Docker secrets --  just thinking about whether there are other scenarios we might want to support in future.